### PR TITLE
Add an integration test for acquireShard

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I added an integration test for the acquireShard method. To do this, I exposed a new `WithFxOptions` `ServerOption` that allows us to modify the constructor graph.

**Why?**
I made this to give us more confidence in the acquireShard method. I also added the `WithFxOption` method to make future testing easier.

**How did you test it?**
With an integration test.

**Potential risks**
Someone could misuse the WithFxOptions option and override dependencies that they really shouldn't.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: temporal/fx.go [hunk 1]
```diff
@@ -132,6 +132,10 @@ type (
 )
 
 func NewServerFx(opts ...ServerOption) *ServerFx {
+	var fxOpts []fx.Option
+	for _, opt := range opts {
+		fxOpts = append(fxOpts, opt.fxOptions()...)
+	}
```
Comment (yiminc) [inline temporal/fx.go:138]: why do we need to make a copy? Can we use opt.fxOptions()... directly to the call of fx.New()

File: temporal/server_option.go [hunk 2]
```diff
@@ -188,3 +191,24 @@ func WithCustomMetricsHandler(provider metrics.MetricsHandler) ServerOption {
...
+// This will override the default config with your custom config.
+// Additionally, you can decorate existing dependencies. For example:
+//
+//	server.WithFxOptions(fx.Decorate(func(cfg *config.Config) *config.Config {
+//		cfg.Persistence.NumHistoryShards = 10
+//		return cfg
+//	}))
+func WithFxOptions(opts ...fx.Option) ServerOption {
+	return fxOptions(opts)
+}
```
Comment (yiminc) [inline temporal/server_option.go:214]: can you follow the same pattern as other server options (return applyFunc)

Comment (MichaelSnowden) [reply temporal/server_option.go:214]: The `applyFunc` is execued while fx is constructing objects, after it constructs the graph. However, we need these options to modify the graph itself, so we have to provide them to fx before that.

File: temporal/fx.go (hunk 1 above)
Comment (MichaelSnowden) [reply temporal/fx.go:138]: Because there could be multiple ServerOptions which provide `fxOptions()` in the `opts` array.

Comment (MichaelSnowden) [reply temporal/fx.go:138]: This is basically flattening a list. It doesn't copy the actual objects, just pointers to them

**Review by yiminc — CHANGES_REQUESTED** [review]
Don't think this is the right way to approach it. Server option and fx are 2 different extension solutions. Server option is currently supported way. Let's talk before you proceed.

**Pull request closed without merging**